pyproject.toml

```diff
@@ -20,11 +20,11 @@ PyJWT = "^2.4.0"
 SQLAlchemy = { extras = ["mypy"], version = "^2.0.25" }
 SQLAlchemy-Utils = "^0.41.1"
 alembic = "^1.13.1"
-dask = { extras = ["dataframe"], version = "^2024.1.1" }
-fastapi = "^0.109.0"
-httpx = { extras = ["http2"], version = "^0.26.0" }
+dask = { extras = ["dataframe"], version = "^2024.4.1" }
+fastapi = "^0.110.0"
+httpx = { version = "0.26", extras = ["http2"] }
 openpyxl = "^3.0.9"
-pandas = "~2.1.2"
+pandas = "^2.2.1"
```

Comment:
Does the fix not work with pandas 2.1? Seems very tight to depend on a package version that was released less than 2 months ago.

Reply:
The initial trigger for this PR was to remove the version pin of pandas. `~2.1.2` translates to `>=2.1.2, <2.2.0`, but climate-processor needs pandas 2.2.0 (if I understand correctly). Updating dependencies in poetry is usually done by running something like `poetry add pandas@latest`, which resolves to the latest version and uses the `^` notation, where `^2.2.1` translates to `>=2.2.1, <3.0.0`. 
In other words, what we have now is simply the default. We could manually lower the requirement to `^2.1.2`, which would allow climate-processor to resolve to the newer versions, too. However, [it looks like](https://github.com/pandas-dev/pandas/commit/4dc3f79c3cfc507204b4b41085d062e8b694472d) pandas is releasing v2.2.2 today, so we won't only support the latest version either way. I think it's fine to keep the dependency as it is right now.

Reply:
I disagree, ixmp4 is now a crucial dependency of pyam and that package is used by many people, so we should be as relaxed as possible - unless there is a good reason to be strict.
See this user feedback https://github.com/IAMconsortium/pyam/issues/840

Reply:
Looking into this.

Reply:
Okay, so I have now used [uv's wonderful `lowest-direct` dependency resolution](https://github.com/astral-sh/uv?tab=readme-ov-file#resolution-strategy) to arrive at a requirements.txt file with version pins for python3.10 that are compatible with our project. This process needs as input a requirements.in file, which I generated from our current pyproject.toml by replacing all version specifiers with `>=`. At first, I was a bit too radical in keeping only the major version; so if we required a project's 2.1.12 version, I'd keep `>= 2.0` in my requirements.in. This produced problems at times and wasn't applicable to projects with versions starting with `0.`. In those cases, I looked at pypi and chose a version more or less at random that is some good six months old.
Next, I created a new venv with python 3.10 and installed all version that were pinned in the requirements.txt (see below). And the tests are successful :)
So I could now hijack this PR and `poetry add` all these dependencies with `>=` marks to expand the dependencies we support somewhat in time. For example, we can still support

- pandas 2.0.0
- dask 2023.9.0
- fastapi 0.100.0

and several others. Please let me know what you think, @danielhuppmann 

<details>
<summary>requirements.txt</summary>

[The versions in here pass our test suite with python 3.10 :)](https://github.com/iiasa/ixmp4/files/14931625/requirements.txt)

</details>

Reply:
@meksor Please also let me know what you think and whether we should revert some more internal dependencies to `^` again. The `dev, docs, server` groups come to mind as groups that might warrant more stringent dependency limits.